src/VisualStudio/Next/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs

```diff
@@ -0,0 +1,163 @@
+            // messages we have for things like registering/unregistering the 
+            // primary solution ID.
+            private readonly object _gate = new object();
+            private Task _lastTask = SpecializedTasks.EmptyTask;
+
+            public WorkspaceHost(VisualStudioWorkspaceImpl workspace)
+            {
+                _workspace = workspace;
+                _currentSolutionId = workspace.CurrentSolution.Id;
+                
+                // Ensure that we populate the remote service with the initial state of
+                // the workspace's solution.
+                EnqueueRegisterPrimarySolutionTask();
+            }
+
+            public void OnAfterWorkingFolderChange()
+            {
+                this.AssertIsForeground();
+                EnqueueRegisterPrimarySolutionTask();
+            }
+
+            public void OnSolutionAdded(SolutionInfo solutionInfo)
+            {
+                this.AssertIsForeground();
+                EnqueueRegisterPrimarySolutionTask();
+            }
+
+            private void EnqueueRegisterPrimarySolutionTask()
+            {
+                _currentSolutionId = _workspace.CurrentSolution.Id;
+                var solutionId = _currentSolutionId;
+
+                lock (_gate)
+                {
+                    _lastTask = _lastTask.ContinueWith(
+                        _ => NotifyRemoteHostOfRegisterPrimarySolution(solutionId),
+                        TaskScheduler.Default).Unwrap();
+                }
+            }
+
+            public void OnBeforeWorkingFolderChange()
+            {
+                this.AssertIsForeground();
+
+                _currentSolutionId = _workspace.CurrentSolution.Id;
+                var solutionId = _currentSolutionId;
+
+                lock (_gate)
+                {
+                    _lastTask = _lastTask.ContinueWith(
+                        _ => NotifyRemoteHostOfUnregisterPrimarySolution(solutionId, synchronousShutdown: true),
+                        TaskScheduler.Default).Unwrap();
+                }
+            }
+
+            public void OnSolutionRemoved()
+            {
+                this.AssertIsForeground();
+
+                // Have to use the cached solution ID we've got as the workspace will
+                // no longer have a solution we can look at.
+                var solutionId = _currentSolutionId;
+                _currentSolutionId = null;
+
+                lock (_gate)
+                {
+                    _lastTask = _lastTask.ContinueWith(
+                        _ => NotifyRemoteHostOfUnregisterPrimarySolution(solutionId, synchronousShutdown: false),
+                        TaskScheduler.Default).Unwrap();
+                }
+            }
+
+            private async Task NotifyRemoteHostOfRegisterPrimarySolution(SolutionId solutionId)
+            {
+                var factory = _workspace.Services.GetService<IRemoteHostClientFactory>();
```

Comment:
I think we create this host inside of RemtoeHostClient creation. so you can just give it to host when you create the host rather than you create one yourself.